Bump AGP to `8.5.1`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR updates Android Gradle Plugin to 8.5.0, mostly needed to address security vulnerabilities.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
1. Install the app in `releaseVanilla` variant as there are changes related to R8.
2. Smoke test the app, focus on testing paths from every module, including `iap` and `cardreader`

### Dependencies
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3065
- https://github.com/wordpress-mobile/WordPress-Android/pull/21082
- https://github.com/WordPress/gutenberg/pull/64044



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Bump AGP to 8.5.0

wzieba committed:
Do not minify libraries, minify app builds only

Starting from Gradle 8.4, shrinking at library level is now respected. Library modules will be now shrinked in isolation, removing code that might be used by other modules. The solution is to not shrink each library module (no need as we don't publish them in isolation) and shrink only app modules.

wzieba committed:
Remove release build type from library modules

They're not used

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Add `POST_NOTIFICATIONS` persmisson to Wear manifest

To address `NotificationPermission` lint error. Actually, this permission is not needed at this moment, as the wear module doesn't use org.wordpress.android.util.AutoForeground or classes that overrides from it so it seems to be a false positive from lint.

I don't want to suppress this issue though, as we might use AutoForeground in the future, and the POST_NOTIFICATIONS will be necessary then.

wzieba committed:
Address `StringFormatMatches` lint warning

- unwrap `String#format(Context#getString` calls. Having two formats is unnecessary and can be confusing
- update format arguments to reflect type declared in strings.xml

wzieba commented on WooCommerce-Wear/src/main/AndroidManifest.xml:
```diff
@@ -5,6 +5,7 @@
 
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="android.permission.WAKE_LOCK" />
+    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
```
More context on this:
<img width="889" alt="Screenshot 2024-07-23 at 11 26 04" src="https://github.com/user-attachments/assets/16edbd5e-dbc2-4018-84be-4045f16c9532">

I think it's a false positive: the `WooCommerce-Wear` doesn't use [AutoForeground](https://github.com/wordpress-mobile/WordPress-Utils-Android/blob/trunk/WordPressUtils/src/main/java/org/wordpress/android/util/AutoForeground.java) or any service that is inheriting from it.

I've experimented with removing `org.wordpress:utils` dependency from `WooCommerce-Wear` module and running lint was resulting with no errors.

To make sure there are no new services, I've compared manifests of WooCommerce-Wear with and without `org.wordpress:utils`. The diff is:
<img width="1255" alt="Screenshot 2024-07-23 at 11 25 15" src="https://github.com/user-attachments/assets/97d11d62-28e0-414c-8458-cf33d93301cf">

which is not relevant to `POST_NOTIFICATIONS`.

My conclusion: Android linter detects `AutoForeground` service in dependencies of the `WooCommerce-Wear` module and throws this lint error, not minding that we don't actually use the service.

I decided to add this permission anyway, as it's already present in `WooCommerce` module and I didn't want to suppress this warning, in case that we'll indeed need posting a notification from the wear app.

wzieba replied:
~Only now I realised that we already request `POST_NOTIFICATIONS` in a few places in the app~
https://github.com/woocommerce/woocommerce-android/blob/e7e9368ab0aa2ed164fd192843eff7dfe60bc109/WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooPermissionUtils.kt#L59-L62
~so adding this `POST_NOTIFICATION` in manifest is more of a correctness (as signilised by lint warning)~

edit: yes, we use but in `WooCommerce` module, not `WooCommerce-Wear`

wzieba commented:
hi @kidinov 👋 can I request your assistance in smoke testing `iap` and `cardreader` modules?

ParaskP7 commented on settings.gradle:
```diff
@@ -1,5 +1,5 @@
 pluginManagement {
-    gradle.ext.agpVersion = '8.1.0'
+    gradle.ext.agpVersion = '8.5.0'
```
Suggestion (💡): A newer version of AGP is now available, the [8.5.1](https://maven.google.com/web/index.html?q=build#com.android.tools.build:gradle:8.5.1), which solve a couple of bugs ([context](https://developer.android.com/studio/releases/fixed-bugs/studio/2024.1.1#android-studio-koala-|-2024.1.1-patch-1)), maybe we should consider updating to that instead? 🤔

wzieba replied:
Sure, thanks! Addressed in 8d45720b9b66eb19a50c621ec913de6b3d3df08d

ParaskP7 replied:
Awesome, thanks @wzieba , I just did another check and everything LGTM!

ParaskP7 reviewed (commented):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected, good job! 🌟 

-----

I have left a single suggestion (💡)  comment for you to consider. I am **NOT** going to approve this PR (for now), leaving the final approval to @kidinov instead ([context](https://github.com/woocommerce/woocommerce-android/pull/12042#issuecomment-2245187584)). No matter, everything worked as expected for me so I can tentatively approve as well if you need me to (just let me know).

-----

`EXTRAS`
1. Warning (⚠️): When this PR get merged into `trunk` dev will need to use AS Koala. AGP 8.5.0 is incompatible with previous versions of AS and thus devs wouldn't be able to (for example) sync the project otherwise (see [Android Studio & AGP compatibility options](https://developer.android.com/studio/releases#android_gradle_plugin_and_android_studio_compatibility)).

wzieba committed:
Bump AGP to 8.5.1

wzieba committed:
Merge branch 'trunk' into agp_8_5_0

wzieba commented:
hi @malinajirka ! 👋 can I have your assistance regarding smoke testing `iap` and `cardreader` modules? The changes introduced with AGP 8.4.0 are related to obfuscating library modules, so I'd like to make sure we don't introduce regression.

I don't have a store with the correct setup, so I think it'll be more efficient if someone from Woo could test it.

malinajirka reviewed (approved):
Hi @wzieba 👋 

> 👋 can I have your assistance regarding smoke testing iap and cardreader modules? The changes introduced with AGP 8.4.0 are related to obfuscating library modules, so I'd like to make sure we don't introduce regression.

I've tested the card reader flows and I haven't noticed any regression.

With the removal of Site Creation in the app, the IAP flows are not used at the moment.

wzieba commented:
Gotcha, thank you @malinajirka ! 🙌

wzieba committed:
Merge branch 'trunk' into agp_8_5_0
